expr/func_substr.go

```diff
@@ -0,0 +1,86 @@
+package expr
+
+import (
+	"strings"
+
+	"github.com/grafana/metrictank/api/models"
+)
+
+type FuncSubstr struct {
+	in    GraphiteFunc
+	start int64
+	stop  int64
+}
+
+func NewSubstr() GraphiteFunc {
+	return &FuncSubstr{}
+}
+
+func NewAliasByMetric() GraphiteFunc {
+	return &FuncSubstr{start: -1, stop: 0}
+}
+
+func (s *FuncSubstr) Signature() ([]Arg, []Arg) {
+	if s.start != 0 || s.stop != 0 {
+		return []Arg{
+			ArgSeriesList{val: &s.in},
+		}, []Arg{ArgSeriesList{}}
+	}
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgInt{val: &s.start, opt: true, key: "start"},
+		ArgInt{val: &s.stop, opt: true, key: "stop"},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncSubstr) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncSubstr) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make([]models.Series, len(series))
```

Comment:
functions that do not affect the datapoints slices (eg that only change the names/tags), don't need to allocate new output series, and don't add them to the cache either (because we currently only use the cache to mark what needs to be returned the pool). i know this is confusing, you may want to freshen up on devdocs/expr.md